File: chroma-league/src/main/java/net/booone/chromaleague/hud/animations/LoseAnimation.java
package net.booone.chromaleague.hud.animations;

import net.booone.razersdk.animation.Animation;
import net.booone.razersdk.animation.Frame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.BreathingColor;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.sdk.ProgressiveRzKeySelector;
import net.booone.razersdk.sdk.RzKey;
import net.booone.razersdk.sdk.RzKeySelector;
import net.booone.razersdk.sdk.json.request.KeyboardEffect;

import java.util.Set;
import java.util.stream.IntStream;

public class LoseAnimation extends Animation {

    private final ProgressiveRzKeySelector progressiveRzKeySelector = new ProgressiveRzKeySelector(
            IntStream.range(0, KeyboardEffect.KEYBOARD_ROWS)
                    .mapToObj(row -> new RzKeySelector().withRow(row).asSet())
                    .toList(), 4);
    private final BreathingColor breathingColor = new BreathingColor(StaticColor.RED, new StaticColor(10, 0, 0), 50);
    private Set<RzKey> currentSet;
    int i = 0;

    @Override
    public Frame getFrame() {
        if (i == 0) {
            currentSet = progressiveRzKeySelector.getNextPart();
            i = 4;
        }
        addToFront(new SimpleFrame(currentSet, breathingColor));
        i--;
        return super.getFrame();
    }
}

File: src/hud/animations/LoseAnimation.ts
import {AnimatedFrame} from '../../razer-sdk/animation/AnimatedFrame.js';
import {SimpleFrame} from '../../razer-sdk/animation/SimpleFrame.js';
import {BreathingColor} from '../../razer-sdk/color/BreathingColor.js';
import {StaticColor} from '../../razer-sdk/color/StaticColor.js';
import {ProgressiveRzKeySelector} from '../../razer-sdk/sdk/ProgressiveRzKeySelector.js';
import {RzKey} from '../../razer-sdk/sdk/RzKey.js';
import {KEYBOARD_ROWS, RzKeySelector} from '../../razer-sdk/sdk/RzKeySelector.js';

function buildRowParts(): ReadonlySet<RzKey>[] {
    const parts: ReadonlySet<RzKey>[] = [];
    for (let row = 0; row < KEYBOARD_ROWS; row++) {
        parts.push(new RzKeySelector().withRow(row).asSet());
    }
    return parts;
}

const TOTAL_FRAMES = 1200; // ~60 seconds at 20fps

export class LoseAnimation extends AnimatedFrame {
    constructor() {
        super();
        const progressiveRzKeySelector = new ProgressiveRzKeySelector(buildRowParts(), 4);
        const breathingColor = new BreathingColor(StaticColor.RED, new StaticColor(10, 0, 0), 50);
        let currentSet: Set<RzKey> = new Set();
        let i = 0;

        for (let frame = 0; frame < TOTAL_FRAMES; frame++) {
            if (i === 0) {
                currentSet = progressiveRzKeySelector.getNextPart();
                i = 4;
            }
            this.addAnimationFrame(new SimpleFrame([...currentSet], breathingColor));
            i--;
        }
    }
}
